pkg/verify/verify.go

```diff
@@ -0,0 +1,204 @@
+		return errors.New("signature on tree head did not verify")
+	}
+
+	// Now verify consistency up to the STH.
+	return ProveConsistency(ctx, rClient, oldSTH, &sth, *logInfo.TreeID)
+}
+
+// VerifyInclusion verifies an entry's inclusion proof. Clients MUST either verify
+// the root hash against a new STH (via VerifyCurrentCheckpoint) or against a
+// trusted, existing STH (via ProveConsistency).
+//nolint
+func VerifyInclusion(ctx context.Context, e *models.LogEntryAnon) error {
+	if e.Verification == nil || e.Verification.InclusionProof == nil {
+		return errors.New("inclusion proof not provided")
+	}
+
+	hashes := [][]byte{}
+	for _, h := range e.Verification.InclusionProof.Hashes {
+		hb, _ := hex.DecodeString(h)
+		hashes = append(hashes, hb)
+	}
+
+	rootHash, err := hex.DecodeString(*e.Verification.InclusionProof.RootHash)
+	if err != nil {
+		return err
+	}
+
+	// Verify the inclusion proof.
+	entryBytes, err := base64.StdEncoding.DecodeString(e.Body.(string))
+	if err != nil {
+		return err
+	}
+	leafHash := rfc6962.DefaultHasher.HashLeaf(entryBytes)
+
+	if err := proof.VerifyInclusion(rfc6962.DefaultHasher, uint64(*e.Verification.InclusionProof.LogIndex),
+		uint64(*e.Verification.InclusionProof.TreeSize), leafHash, hashes, rootHash); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// VerifySignedEntryTimestamp verifies the entry's SET against the provided
+// public key.
+//nolint
+func VerifySignedEntryTimestamp(ctx context.Context, e *models.LogEntryAnon, verifier signature.Verifier) error {
+	if e.Verification == nil {
+		return fmt.Errorf("missing verification")
+	}
+	if e.Verification.SignedEntryTimestamp == nil {
+		return fmt.Errorf("signature missing")
+	}
+
+	type bundle struct {
+		Body           interface{} `json:"body"`
+		IntegratedTime int64       `json:"integratedTime"`
+		// Note that this is the virtual index.
+		LogIndex int64  `json:"logIndex"`
+		LogID    string `json:"logID"`
+	}
+	bundlePayload := bundle{
+		Body:           e.Body,
+		IntegratedTime: *e.IntegratedTime,
+		LogIndex:       *e.LogIndex,
+		LogID:          *e.LogID,
+	}
+	contents, err := json.Marshal(bundlePayload)
+	if err != nil {
+		return fmt.Errorf("marshaling bundle: %w", err)
+	}
+	canonicalized, err := jsoncanonicalizer.Transform(contents)
+	if err != nil {
+		return fmt.Errorf("canonicalizing bundle: %w", err)
+	}
+
+	// verify the SET against the public key
+	if err := verifier.VerifySignature(bytes.NewReader(e.Verification.SignedEntryTimestamp),
+		bytes.NewReader(canonicalized), options.WithContext(ctx)); err != nil {
+		return fmt.Errorf("unable to verify bundle: %w", err)
+	}
+	return nil
+}
+
+// VerifyLogEntry performs verification of a LogEntry given a Rekor verifier.
+// Performs inclusion proof verification up to a verified root hash and
+// SignedEntryTimestamp verification.
+//nolint
+func VerifyLogEntry(ctx context.Context, rClient *client.Rekor, e *models.LogEntryAnon, verifier signature.Verifier) error {
```

Comment:
Should this take a `rClient`? I think we'll need a function that is solely for offline verification (Inclusion, STH signature, SET), and one for online (Inclusion, signature, consistency, maybe SET too?)

Reply:
I can remove now that we omit STH verification.

Reply:
The one that also verifies consistency can be added as a separate function, agreed, like `VerifyLogEntryWithConsistency`